### PR TITLE
[FEATURE: UIPT-40]  - Filter

### PR DESCRIPTION
**Pull request by TorskyM**

### CODE REVIEW

File: src/app/app.enum.ts [hunk 1]
```diff
@@ -25,6 +25,9 @@ export enum ToastrMessage {
   internalError = 'An internal error',
   emptyWishList = 'Your Wishlist is currently empty',
   productAddedToCart = 'Product has been added to card',
+  paymentSeccessfull = 'Payment successful!',
```
Comment (castan1c) [inline src/app/app.enum.ts]: paymentSeccessfull typo

File: src/app/shared/product-filter/product-filter.component.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    this.products.forEach(item => {
+      this.productCategory.push(item.category);
+      this.productBrand.push(item.brand);
+    });
+    this.productCategory = [...new Set(this.productCategory)];
+    this.productBrand = [...new Set(this.productBrand)];
+  }
+
+  private onFormatItem(unformatedItems) {
```
Comment (castan1c) [inline src/app/shared/product-filter/product-filter.component.ts]: Could you please specify return type for this method?

File: src/app/shared/product-filter/product-filter.component.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      this.selectedProducts = [...new Set(this.selectedProducts)];
+
+    if (!this.selectedProducts.length) {
+      this.toastrService.warning(ToastrMessage.filterFail);
+    }
+
+    this.dataChange.emit(this.onFormatItem(this.selectedProducts));
+  }
+
+  private onStrictRequest(selectedProducts, strictRequest: Array<string>) {
```
Comment (castan1c) [inline src/app/shared/product-filter/product-filter.component.ts]: Same, return type

File: src/app/shared/product-filter/product-filter.scss [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+::-webkit-scrollbar-thumb {
+  background: #888;
+}
+
+::-webkit-scrollbar-thumb:hover {
+  background: #555;
+}
+
+.filter-wrapper {
```
Comment (castan1c) [inline src/app/shared/product-filter/product-filter.scss]: Could you rewrite it like this: 
`.filter {
  &-wrapper {
    ...
  }
}`

File: src/app/shared/product-filter/product-filter.scss [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
+  background-color: $bgColorMedium;
+}
+
+.filter-container {
```
Comment (castan1c) [inline src/app/shared/product-filter/product-filter.scss]: `&-container`

File: src/app/shared/product-filter/product-filter.scss [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  }
+
+  &__collapse {
+    margin: 30px 0 5px 0;
+    box-shadow: 0 0 10px rgba(33, 33, 33, 0.2);
+    background-color: $bgColorLight;
+  }
+}
+
+.filter-criteria {
```
Comment (castan1c) [inline src/app/shared/product-filter/product-filter.scss]: `&-criteria`

File: src/app/shared/product-filter/product-filter.scss [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+  &__subtitle {
+    display: block;
+    text-align: left;
+    margin-left: 30px;
+    font-size: 0.9rem;
+  }
+}
+
+.filter-button {
```
Comment (castan1c) [inline src/app/shared/product-filter/product-filter.scss]: `&-button`

File: src/app/shared/shop-by-category/shop-by-category.html [hunk 8]
```diff
@@ -1,18 +1,26 @@
...
-      <app-product-item *ngFor="let product of filteredItems$ | async"
-        class="product__item"
+      <app-product-item
+        *ngFor="let product of filteredItems"
         [status]="product.status"
         [imgUrl]="product.imgUrl"
         [imgUrlNext]="product.imgUrlNext"
         [productTitle]="product.productTitle"
         [productPrice]="product.productPrice"
-        [productId]="product.productId">
```
Comment (castan1c) [inline src/app/shared/shop-by-category/shop-by-category.html]: Could you please explain why do we need to pass all this attributes to product-item?

File: src/app/shared/shop-by-category/shop-by-category.component.ts [hunk 9]
```diff
@@ -37,18 +37,22 @@ export class ShopByCategoryComponent implements OnInit, OnDestroy {
...
 
+  public onDataChange(event): void {
+    this.filteredItems = event;
+  }
+
   public ngOnDestroy(): void { }
 
   private getProductsByCategory(): void {
-    this.filteredItems$ = this.store
+    this.store
```
Comment (castan1c) [inline src/app/shared/shop-by-category/shop-by-category.component.ts]: You changed stream to subscription, but forot to add Subscription and AutoUnsubscribe()

File: src/app/shared/product-filter/product-filter.scss [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+      width: 5px;
+    }
+  }
+
+  &:hover {
+    border: 5px solid $shadowStyle;
+  }
+}
+
+.label__checkbox:checked + .label__text .label__check {
```
Comment (castan1c) [inline src/app/shared/product-filter/product-filter.scss]: `.label {
  &__checbox:checked + &__text &__check {
  }
}` should be the same

File: src/app/shared/shop-by-category/shop-by-category.html (hunk 8 above)
Comment (TorskyM) [reply src/app/shared/shop-by-category/shop-by-category.html]: I agree, we can just transfer the product itself to the component and take data directly from it, but we still need to set the initial data
Shall I do this now?

Comment (castan1c) [reply src/app/shared/shop-by-category/shop-by-category.html]: Nvm, i just realized the product-item-component had those inputs all the time and it is used ike that in the whole project

File: src/app/interfaces/product-media.interface.ts [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import { IProductImage } from '.';
```
Comment (IrynaMukhina) [inline src/app/interfaces/product-media.interface.ts]: is this import needed here?

File: src/app/shared/header/main-menu/main-menu.html [hunk 12]
```diff
@@ -1,5 +1,5 @@
 <div class="menu">
-  
+
```
Comment (IrynaMukhina) [inline src/app/shared/header/main-menu/main-menu.html]: please remove this space

File: src/app/shared/product-filter/product-filter.component.ts [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+  selector: 'app-product-filter',
+  templateUrl: './product-filter.html',
+  styleUrls: ['./product-filter.scss']
+})
+
+export class ProductFilterComponent implements OnInit, OnDestroy {
+  constructor(
+    private store: Store<IAppState>,
+    private productService: ProductService,
+    private toastrService: ToastrService) { }
```
Comment (IrynaMukhina) [inline src/app/shared/product-filter/product-filter.component.ts]: please align

File: src/app/shared/product-filter/product-filter.component.ts [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+        this.strictRequest.category.push(this.criteriaName) :
+        this.strictRequest.brand.push(this.criteriaName);
+    } else {
+      indexCategory !== -1 ?
+        this.strictRequest.category = this.strictRequest.category
+          .filter(criteria => criteria !== this.criteriaName) :
+        this.strictRequest.brand = this.strictRequest.brand
+          .filter(criteria => criteria !== this.criteriaName);
+    }
+    this.onFilter();
```
Comment (IrynaMukhina) [inline src/app/shared/product-filter/product-filter.component.ts]: please add an empty line

File: src/app/shared/product-filter/product-filter.component.ts [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+  private onSetFilters(): void {
+    this.productCategory = [];
+    this.productBrand = [];
+
+    this.products.forEach(item => {
+      this.productCategory.push(item.category);
+      this.productBrand.push(item.brand);
+    });
+    this.productCategory = [...new Set(this.productCategory)];
```
Comment (IrynaMukhina) [inline src/app/shared/product-filter/product-filter.component.ts]: empty line before

File: src/app/shared/product-filter/product-filter.component.ts [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        this.products.forEach(product => {
+          if (product.category === category) {
+            this.selectedProducts.push(product);
+
+            this.onStrictRequest(this.selectedProducts, this.strictRequest.brand);
+          }
+        })
+      );
+    }
+    this.strictRequestItems.length ?
```
Comment (IrynaMukhina) [inline src/app/shared/product-filter/product-filter.component.ts]: empty line before

File: src/app/shared/product-filter/product-filter.html [hunk 17]
```diff
@@ -0,0 +1,48 @@
+<div class="filter-wrapper">
+<div class="filter-container">
+  <div class="filter-container__align">
+  <h2 class="filter-criteria__title">Brand</h2>
+  <div class="filter-criteria__wrapper">
+    <div class="filter-criteria" *ngFor="let product of productBrand">
+        <div class="center">
+          <label class="label">
+      <input class="filter-criteria__option label__checkbox" (change)="onCreateRequest($event)" type="checkbox" name="brand"
```
Comment (IrynaMukhina) [inline src/app/shared/product-filter/product-filter.html]: please align

File: src/app/shared/product-filter/product-filter.html [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    </div>
+  </div>
+</div>
+  <div class="filter-container__collapse">
+  <h2 class="filter-criteria__title">Category</h2>
+  <div class="filter-criteria__wrapper">
+    <div class="filter-criteria" *ngFor="let product of productCategory">
+        <div class="center">
+            <label class="label">
+      <input class="filter-criteria__option label__checkbox" (change)="onCreateRequest($event)" type="checkbox" name="category"
```
Comment (IrynaMukhina) [inline src/app/shared/product-filter/product-filter.html]: please align too

File: src/app/shared/product-filter/product-filter.scss [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+
+:host {
+  margin: 0 auto 0 20px;
+}
+
+::-webkit-scrollbar {
+  width: 7px;
+}
+
+::-webkit-scrollbar-track {
```
Comment (IrynaMukhina) [inline src/app/shared/product-filter/product-filter.scss]: can you use here such way?
::-webkit-scrollbar {
&-track { }
}

File: src/app/shared/product-filter/product-filter.scss [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+  background: $bgColorLight;
+}
+
+::-webkit-scrollbar-thumb {
+  background: #888;
+}
+
+::-webkit-scrollbar-thumb:hover {
```
Comment (IrynaMukhina) [inline src/app/shared/product-filter/product-filter.scss]: can you use nesting?

**Comment by IrynaMukhina** [comment]
@TorskyM could you please resolve conflict?

File: src/app/interfaces/product-media.interface.ts (hunk 11 above)
Comment (TorskyM) [reply src/app/interfaces/product-media.interface.ts]: I see this for the first time)

**Pull request merged**